apps/challenges/aws_utils.py

```diff
@@ -1487,3 +1584,287 @@ def challenge_approval_callback(sender, instance, field_name, **kwargs):
+    calls the delete_eks_cluster function
+
+    Arguments:
+        instance {<class 'django.db.models.query.QuerySet'>} -- instance of the model
+    """
+
+    from .models import ChallengeEvaluationCluster
+    from .serializers import ChallengeEvaluationClusterSerializer
+    from .utils import get_aws_credentials_for_challenge
+
+    for obj in serializers.deserialize("json", challenge):
+        challenge_obj = obj.object
+    challenge_evaluation_cluster = (
+        ChallengeEvaluationCluster.objects.get(challenge=challenge_obj)
+    )
+    if challenge_obj.approved_by_admin and challenge_obj.is_docker_based:
+        challenge_aws_keys = get_aws_credentials_for_challenge(
+            challenge_obj.pk
+        )
+        client = get_boto3_client("eks", challenge_aws_keys)
+        try:
+            # delete the nodegroup
+            response = client.delete_nodegroup(
+                clusterName=challenge_evaluation_cluster.name,
+                nodegroupName=challenge_evaluation_cluster.nodegroup_name
+            )
+            waiter = client.get_waiter('nodegroup_deleted')
+            waiter.wait(
+                clusterName=challenge_evaluation_cluster.name,
+                nodegroupName=challenge_evaluation_cluster.nodegroup_name
+            )
+            # Deleting the cluster after nodegroup cluster is deleted
+            delete_eks_cluster.delay(challenge)
+            
+            return response
+        except ClientError as e:
+            logger.exception(e)
+            return
+        
+
+
+@app.task    
+def delete_eks_vpc(challenge):
+    """
+    Remove VPC and associated Subnets, InternetGateways, RouteTables and SecurityGroups from AWS
+    calls delete_eks_nodegroup function
+
+    Arguments:
+        instance {<class 'django.db.models.query.QuerySet'>} -- instance of the model
+    """
+    from .models import ChallengeEvaluationCluster
+    from .serializers import ChallengeEvaluationClusterSerializer
+    from .utils import get_aws_credentials_for_challenge
+
+    for obj in serializers.deserialize("json", challenge):
+        challenge_obj = obj.object
+    challenge_evaluation_cluster = (
+        ChallengeEvaluationCluster.objects.get(challenge=challenge_obj)
+    )
+    vpcid = challenge_evaluation_cluster.vpc_id
+    if not vpcid:
+        return
+    challenge_aws_keys = get_aws_credentials_for_challenge(
+        challenge.pk
+    )
+    ec2 = get_boto3_resource("ec2", challenge_aws_keys)
+    client = ec2.meta.client
+    
+    vpc = ec2.Vpc(vpcid)
+    # detach default dhcp_options if associated with the vpc
+    dhcp_options_default = ec2.DhcpOptions('default')
+    if dhcp_options_default:
+        dhcp_options_default.associate_with_vpc(
+            VpcId=vpc.id
+        )
+    # detach and delete all gateways associated with the vpc
+    for gw in vpc.internet_gateways.all():
+        try:
+            vpc.detach_internet_gateway(InternetGatewayId=gw.id)
```

Comment:
Deleting VPC should delete all attached resources. If that doesn't work please using APIs like `delete_vpc` `delete_subnet` `delete_internet_gateway`

Reply:
The `internet_gateway` must be detached from the VPC before it can be deleted as mentioned [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.delete_internet_gateway)